### ISSUE TITLE
Bindings to CSSStyleDeclaration object are buggy

### ISSUE THREAD

**Issue by jayeshbhoot** [darklang/rescript-tea#30]
`style` type, probably representing the `CSSStyleDeclaration` object in JS, is defined as following in https://github.com/pbiggar/rescript-tea/blob/master/src/web_node.res#L1:

```
type style = {
  @get
  "setProperty": Js.undefined<Web_json.t> /* TODO:  Revamp this and the next line... */,
  @meth
  "setProperty__": (string, Js.null<string>, Js.null<string>) => unit,
}

@send external
setProperty__: (string, Js.null<string>, Js.null<string>) => unit="setProperty__"
```

## Problems

I am not sure what purpose `@get "setProperty"` and `@meth "setProperty__"` inside the `style` type serve, so I will only focus on `@send external setProperty__` binding.

1. `setProperty__` does not exist on the JS object `CSSStyleDeclaration`.
2. `style` type is missing as the first argument in the `@send external setProperty__` definition.
3. As a result of 2, `@send external` is defined on `string` type, not `style` type.
4. This erroneous definition is used here (https://github.com/pbiggar/rescript-tea/blob/master/src/web_node.res#L83), where `setProperty__` is indeed called on `key`, not `style` object.

All of the above result into a **runtime** error: `TypeError: key.setProperty__ is not a function. (In 'key.setProperty__(value3, priority ? "important" : null)', 'key.setProperty__' is undefined)`

## Fix

Fix the binding to

```
// Get rid of `@get "setProperty"` and `@meth "setProperty__"` IF they are not used anywhere.
type style

@send external
setProperty: (style, string, Js.null<string>, Js.null<string>) => unit="setProperty"
```

Then use it as

```
style->setProperty(
      key,
      value,
      if priority {
        Js.Null.return("important")
      } else {
        Js.Null.empty
      },
    )
```

We could improve on the above as:

```
module CSSStyleDeclaration = {
  type t


@send external
setProperty: (t, string, Js.null<string>, Js.null<string>) => unit="setProperty"
}
```

but I think your top priority is a 1-to-1 ocaml-to-rescript conversion.

If you are open to accept PRs, I could make one in the coming week.

**Comment by pbiggar** [comment]
Thanks! I'm actually working on this at the moment. I've been trying to update the darklang client to use the one in the master branch here, and I think it's partially broken.

However, my plan is to remove most of the stuff in the `Web` module, and use bindings from either the [Dom](https://rescript-lang.org/docs/manual/latest/api/dom) module or [rescript-webapi](https://github.com/tinymce/rescript-webapi)

**Comment by jayeshbhoot** [comment]
> I've been trying to update the darklang client to use the one in the master branch here, and I think it's partially broken.

Yes. I noticed that after I submitted this issue. The `rescript-tea` dependency in the darklang repo is still pinned to the last commit of the original OvermindDL1/bucklescript-tea repo.

During my own experimentation, I found that the `rescript convert` feature leaves a lot of bugs in the converted code. For example, similar to the buggy `setProperty__` binding, the generated code (https://github.com/pbiggar/rescript-tea/blob/master/src/web_xmlhttprequest.res#L72) contains a buggy `xhr._open()` binding too. `_open()` does not exist in JS XHR API, but `open()` does.

As you are still poking around in this repo, I decided to fork the original OvermindDL1/bucklescript-tea repo and adapt it to my needs for now.

In this fork (https://github.com/jayeshbhoot/rescript-tea), I have chosen to not convert to rescript code right away. Instead, I will go through the following steps.

1. Remove the ocaml version. While I prefer ocaml to reason, reason is syntactically more similar to rescript, which should come in handy in the final step.
2. Incrementally remove all of the unnecessary cruft, like `Tea_result`.
3. Similar to your plan, get rid of the `Web` module. Replace its types with the built-in `Dom` module + custom-defined external bindings. I won't use `rescript-webapi` for now.
4. Replace the use of `Pervasives` module with `Js` and `Belt` modules.
5. Only after the bare minimum `reasonml` code is left, will I attempt to convert it to rescript - manually if needed.

I have finished the first 2 steps. The project builds, all the tests  successfully, another project (https://github.com/jayeshbhoot/rescript-tea-poc) that depends on this one – one also runs successfully.

**Comment by pbiggar** [comment]
Looks like we're doing a lot of the same work. Here's my progress: https://github.com/pbiggar/rescript-tea/pull/32

I abandoned the current master branch for now, as I work out where we are. My current goal is to separately get the darklang and rescript-tea repos using rescript 9.1.4 - I don't think cross-version dependencies are going to work here, at least while the code is bucklescript/reason.

Similarly to you, I decided not to convert to rescript initially, but instead to try and remove as much of Web_* as possible first, so that there's less to mess up.

**Comment by pbiggar** [comment]
I think you were probably right to keep the reason code instead of the ocaml code. I kept the ocaml code because the reason code was already autogenerated from the ocaml, which was AFAIK the preferred language of the original author, but I'm not sure if that's going to be a win or a loss.

The module I'm having the most struggle with is web_xmlhttprequest. If you manage to get it upgraded I'd be delighted to copy it :)

**Comment by pbiggar** [comment]
My planned steps:
- [x] remove dual syntax
- [x] remove tea_result
- [ ] Remove as many of the web_* modules as possible, using Webapi-rescript instead
  - [x] date
  - [ ] window (half-done)
  - [x] location
  - [x] event
  - [x] node
  - [ ] formdata
  - [ ] json
  - [ ] window_history
  - [ ] window_localstorage (half-done)
- [ ] convert remaining code to rescript, a file at a time
  - [x] vdom
- [ ] harvest other changes from master branch
  - [ ] convert all names to camelCase https://github.com/pbiggar/rescript-tea/commit/bb73e8ff841c284c09e88662b187c7294947e33a https://github.com/pbiggar/rescript-tea/commit/b0eba356bdc4369e9d103e912f286870748441e2
  - [ ] add in circleci config https://github.com/pbiggar/rescript-tea/commit/9db54ce9641c20d6b250fb6a25a373da9e61c490
  - [ ] remove apostrophes https://github.com/pbiggar/rescript-tea/commit/45a1a1dec7b30b0a37e9f9dc0e46edfe6bdeef04
  - [ ] add test
  - [ ] fix spellcheck https://github.com/pbiggar/rescript-tea/commit/64000896164dc134a7f7baba098205e713b42fd3
  - [ ] replace Tea_html with Tea_html2 https://github.com/pbiggar/rescript-tea/commit/2b6063988c41c77b682cea054aaea9833d1b70d0
  - [ ] ensure keys are labels and present on all events https://github.com/pbiggar/rescript-tea/commit/14aad4a92d519824d30505af2e5f4a04d070db3f

**Comment by pbiggar** [comment]
Needless to say, it would be great if we could find a way to work together on this - we're taking similar approaches, but currently our work would be challenging to integrate due to the ocaml/reason differences.

**Comment by pbiggar** [comment]
@jayeshbhoot I think my branch is has everything working now, and converted to rescript syntax. Could you take a look? https://github.com/pbiggar/rescript-tea/pull/32

**Comment by jayeshbhoot** [comment]
@pbiggar Dropping a comment to acknowledge all of your comments. 

> I think you were probably right to keep the reason code instead of the ocaml code. I kept the ocaml code because the reason code was already autogenerated from the ocaml, which was AFAIK the preferred language of the original author, but I'm not sure if that's going to be a win or a loss.

Not everything is rosy on reason side.

> reason code was already autogenerated from the ocaml, which was AFAIK the preferred language of the original author

And it shows. One function was entirely missing in reasonml module (probably this one was hand-written instead of auto-generated).

> Needless to say, it would be great if we could find a way to work together on this - we're taking similar approaches, but currently our work would be challenging to integrate due to the ocaml/reason differences.

Its likely that I will defer to your fork sooner or later anyway. I don't mind the OCaml-to-ReScript approach, but I am reluctant to swell the dependency graph by adding rescript-webapi, which is quite bulky. But who knows, this is my inexperience and not-yet-battered motivation speaking!

> The module I'm having the most struggle with is web_xmlhttprequest. If you manage to get it upgraded I'd be delighted to copy it :)

> I think my branch is has everything working now, and converted to rescript syntax. Could you take a look?

These days I am racing to meet a deadline (which had compelled me to quickly adapt bs-tea to my needs in the first place). I should be able to have a look on this weekend.

P.S.: Thank you for all the efforts!

**Comment by pbiggar** [comment]
The redone merge is now merged into the main branch. A review would still be very welcome though (or testing it out on your stuff - it works with darklang code)

**Comment by pbiggar** [comment]
I've pulled all the extra features that were in the old main branch, and confirmed they all work on the darklang codebase. So gonna call this done. Let me know if there are any issues and we can get them solved.

**Comment by jayeshbhoot** [comment]
I finally got time today to switch to your repo.

The transition went smooth. I only had to fix the type errors caused by the switch to rescript-webapi.

Everything works. Great work!

Looking forward to contribute to this repo.
